### PR TITLE
Dump dpll pin's states 

### PR DESCRIPTION
**Pull request by gtannous-spec**
## Summary by CodeRabbit

* **Bug Fixes**
  * Adjusted pin configuration read-back to confirm only via raw pin-get results, avoiding human-readable formatting/logging on success.
  * Improved DPLL update diagnostics by logging interface/type and the computed lock-state name.
* **New Features**
  * Updated pin-table reporting to show only relevant input pins per DPLL and clock, with lock-state-aware parent selection and compact pin/parent state columns.
* **Tests**
  * Added unit tests validating lock-state and parent filtering logic, clock/device filtering, label rendering, and edge cases (nil/empty pins, nil priority).

### CODE REVIEW

File: pkg/event/event_tbc.go [hunk 1]
```diff
@@ -560,6 +562,11 @@ func (e *EventHandler) getLargestOffset(cfgName string) int64 {
 		}
 	}
 	glog.Info("Largest offset ", worstOffset)
+	if cs, ok := e.clkSyncState[cfgName]; ok && cs.state == PTP_HOLDOVER &&
+		worstOffset != FaultyPhaseOffset &&
+		int64(math.Abs(float64(worstOffset))) > int64(math.Abs(float64(cs.maxHoldoverOffset))) {
+		cs.maxHoldoverOffset = worstOffset
+	}
```
[the remark this answers is not in the document]
Comment (gtannous-spec) [reply pkg/event/event_tbc.go]: `getLargestOffset()` is intentionally used because it captures the ptp4l-level offset.
`getCalculatedHoldoverOffset()` returns the DPLL's internal phase offset (typically 0-9ns), which is what gets reported in the T-BC-STATUS log line. Switching to it would make maxHoldoverOffset always show ~0-9ns and would be hiding the actual drift the feature was designed to surface.

File: pkg/daemon/daemon.go [hunk 2]
```diff
@@ -1365,6 +1365,15 @@ func (dn *Daemon) applyNodePtpProfile(runID int, nodeProfile *ptpv1.PtpProfile)
...
 					})
+					np := nodeProfile
+					dpllDaemon.OnSourceLostChanged = func(iface string, sourceLost bool) {
+						if sourceLost {
+							glog.Infof("DPLL %s source lost: requesting pin dump", iface)
+						} else {
+							glog.Infof("DPLL %s source acquired: requesting pin dump", iface)
+						}
+						dn.pluginManager.AfterRunPTPCommand(np, "dpll-pin-dump")
+					}
```
[the remark this answers is not in the document]
Comment (gtannous-spec) [reply pkg/daemon/daemon.go]: Fixed.

File: pkg/event/event_tbc.go [hunk 3]
```diff
@@ -153,6 +153,7 @@ func (e *EventHandler) updateBCState(event Event) (clockSyncState, bool, bool) {
 		} else if e.isSourceLostBC(cfgName) {
 			e.clkSyncState[cfgName].state = PTP_HOLDOVER
 			e.clkSyncState[cfgName].clockClass = fbprotocol.ClockClass(135)
+			e.clkSyncState[cfgName].maxHoldoverOffset = 0
```
Comment (vitus133) [inline pkg/event/event_tbc.go]: Why is this change?

File: pkg/event/event_tbc.go (hunk 1 above)
Comment (vitus133) [inline pkg/event/event_tbc.go]: Largest offset is already logged
```
I0629 07:51:13.404111 3216083 event_tbc.go:564] Largest offset 2

```

**Review by vitus133 — CHANGES_REQUESTED** [review]
Pin states are already logged when we change them on exactly the same events:
```
I0629 07:48:11.347208 3216083 clock-chain.go:634] pin reply: {"timestamp":"2026-06-29T07:48:11.347157889Z","id":52,"moduleName":"ice","clockId":"0x507c6fffff0ac7e0","boardLabel":"GNSS-1PPS","type":"gnss","frequency":1,"frequencySupported":[{"frequencyMin":1,"frequencyMax":1}],"capabilities":"state-can-change,priority-can-change","pinParentDevice":[{"parentID":6,"direction":"input","prio":0,"state":"selectable","phaseOffsetPs":0,"operstate":"unknown"},{"parentID":7,"direction":"input","prio":0,"state":"selectable","phaseOffsetPs":0,"operstate":"unknown"}],"phaseAdjustMin":-100000000,"phaseAdjustMax":100000000,"phaseAdjust":0,"operstate":"unknown"}
```
DPLL state changes are also captured in the log.
A much cleaner approach would be to print a list of input pins with their settings on every DPLL notification. The notifications are sent when DPLL device changes its state, so it is exactly what you wanted IIUC. Check around line 386 or 437 for a potential place to implement it.

When you print the pins, it could be a lot with 3 DPLLs in the system. I suggest to print just the ones with the administrative state selectable or active on any of the parents. All others are irrelevant, so you don't need to hardcode their names. Just dump all pins and filter by the state, whatever label comes. This way you also make it  hardware-independent and plugin independent.
If you do a table format pin print() method, please replace in all other places where it is now JSON.  Include a hexadecimal clock ID, pin ID, prio for all parents (one or many) and administrative and operational state. For SyncE there could be a lot of parents, so it can become ugly, be aware of it:
```
pin id 25:
  module-name: zl3073x
  clock-id: 3549816820761526005
  board-label: 156M25_NAC0_CLKREF_SYNC
  package-label: OUT3
  type: synce-eth-port
  capabilities: 0x0
  phase-adjust-min: -2147483648 ps
  phase-adjust-max: 2147483647 ps
  phase-adjust-gran: 800
  parent-device:
    id 0 direction input state disconnected
    id 1 direction input state disconnected
    id 2 direction input state disconnected
    id 3 direction input state disconnected
    id 4 direction input state disconnected
    id 5 direction input state disconnected
    id 6 direction input state disconnected
    id 7 direction input state disconnected
    id 10 direction output state connected
```

File: pkg/event/event_tbc.go (hunk 3 above)
Comment (gtannous-spec) [reply pkg/event/event_tbc.go]: maxHoldOver offset was added to track the highest jump in the offset during HOLDOVER, so once the FSM goes from locked to holdover, it keeps tracking maxHoldOverOffset reached during holdover. 
It is not crucial, just helped me to track highest offset as it was jumping to 40-60~ ns sometimes during holdover

File: pkg/event/event_tbc.go (hunk 1 above)
Comment (gtannous-spec) [reply pkg/event/event_tbc.go]: Largest offset tracks the largest offset in a given window, not in the whole HOLDOVER state, 
So I wanted to have the maximum of the candidates that were chosen to be the largest in their offset windows

Comment (gtannous-spec) [reply pkg/event/event_tbc.go]: I put it bcuz it helped me a bit while debugging, in any case we can drop it easily at any point, that's why I put it in a separate commit :)

**Comment by gtannous-spec** [comment]
I chose to dump the pins state of the DPLL on the DUT, while I know that it is sort of captured on transitions but it is not clear at all, this way we have a cleaner illustration into Pin names/states/priorities upon transition instead of having merely the id.

> Just dump all pins and filter by the state, whatever label comes
Also I didn't understand the filter by state, do you mean we shall print only the ones are connected?

@vitus133

File: addons/intel/clock-chain.go [hunk 4]
```diff
@@ -445,6 +446,38 @@ func (c *ClockChain) InitPinsTBC() error {
...
+			}
+			glog.Infof("  pin %d %-14s parentID=%d dir=%-6s prio=%-4s state=%-12s operstate=%s",
+				pin.ID, pin.BoardLabel, pd.ParentID,
+				dpll.GetPinDirection(pd.Direction),
+				prioStr,
+				dpll.GetPinState(pd.State),
+				dpll.GetPinOperstate(pd.Operstate))
+		}
+	}
+	glog.Infof("=== end DPLL pin dump ===")
```
Comment (nocturnalastro) [inline addons/intel/clock-chain.go]: I think these should have a >10 verbosity as these are debug prints, not sure we want them in all logs.

File: pkg/event/event_tbc.go (hunk 1 above)
Comment (nocturnalastro) [reply pkg/event/event_tbc.go]: In holdover the offset increases on a calculated ramp? So why wouldn't that just be the last value in the window?

Comment (gtannous-spec) [reply pkg/event/event_tbc.go]: It starts to descend at some point

Comment (gtannous-spec) [reply pkg/event/event_tbc.go]: Using it for personal purposes ! got rid of the commit.

File: addons/intel/clock-chain.go (hunk 4 above)
Comment (gtannous-spec) [reply addons/intel/clock-chain.go]: We do want to log the artifacts upon every DPLL notification ! that will help a lot later with the hardwareconfig path transitions

File: addons/intel/clock-chain.go [hunk 5]
```diff
@@ -626,12 +625,16 @@ func batchPinSet(commands []dpll.PinParentDeviceCtl) error {
...
-			return err
+		for _, pd := range info.ParentDevice {
+			prioStr := "n/a"
+			if pd.Prio != nil {
+				prioStr = fmt.Sprintf("%d", *pd.Prio)
+			}
+			glog.Infof("  pin=%-3d %-14s clockID=0x%x parentID=%-2d dir=%-6s prio=%-4s admin=%-12s oper=%s",
+				info.ID, info.BoardLabel, info.ClockID, pd.ParentID,
+				dpll.GetPinDirection(pd.Direction), prioStr,
+				dpll.GetPinState(pd.State), dpll.GetPinOperstate(pd.Operstate))
```
Comment (vitus133) [inline addons/intel/clock-chain.go]: Does it need to be duplicated here and in the hardwareconfig path? Maybe a short helper function accessible from both places would help?

Comment (vitus133) [reply addons/intel/clock-chain.go]: Also, please consider adding the package label when not empty. It is very helpful, as Dell and HPE have different board labels, but same package label

Comment (gtannous-spec) [reply addons/intel/clock-chain.go]: Fixed !

File: pkg/dpll-netlink/pin_table_test.go [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+	assert.Equal(t, 2, len(groups[0].lines), "0xAABB has 2 pins")
+	assert.Equal(t, uint64(0xCCDD), groups[1].clockID)
+	assert.Equal(t, 1, len(groups[1].lines), "0xCCDD has 1 pin")
+
+	assert.Contains(t, groups[0].lines[0], "CVL-SDP22")
+	assert.Contains(t, groups[0].lines[0], "prio=255")
+	assert.Contains(t, groups[0].lines[1], "CVL-SDP23")
+	assert.Contains(t, groups[0].lines[1], "admin=connected")
+	assert.Contains(t, groups[1].lines[0], "prio=5")
+}
```
[the remark this answers is not in the document]
Comment (gtannous-spec) [reply pkg/dpll-netlink/pin_table_test.go]: Fixed.

**Review by vitus133 — COMMENTED** [review]
Yep, looks good. Could you please send a log example with these prints?

**Comment by gtannous-spec** [comment]
This is an example of what logs show when a DPLL event happen (FREERUN-LOCKED): @vitus133 
```
I0707 09:14:09.343265 3202930 pin_table.go:35] === DPLL pin table (ens2f0 pps->LOCKED) clockID=0x507c6fffff4b9950 ===
  pin=0   CVL-SDP22            parentID=0  dir=input  prio=255  admin=selectable   oper=unknown
  pin=0   CVL-SDP22            parentID=1  dir=input  prio=5    admin=selectable   oper=unknown
  pin=1   CVL-SDP20            parentID=0  dir=input  prio=255  admin=selectable   oper=unknown
  pin=1   CVL-SDP20            parentID=1  dir=input  prio=4    admin=selectable   oper=unknown
  pin=2   C827_0-RCLKA         parentID=0  dir=input  prio=8    admin=selectable   oper=unknown
  pin=2   C827_0-RCLKA         parentID=1  dir=input  prio=8    admin=selectable   oper=unknown
  pin=3   C827_0-RCLKB         parentID=0  dir=input  prio=9    admin=selectable   oper=unknown
  pin=3   C827_0-RCLKB         parentID=1  dir=input  prio=9    admin=selectable   oper=unknown
  pin=4   SMA1                 parentID=0  dir=input  prio=3    admin=selectable   oper=unknown
  pin=4   SMA1                 parentID=1  dir=input  prio=3    admin=selectable   oper=unknown
  pin=5   SMA2/U.FL2           parentID=0  dir=input  prio=2    admin=selectable   oper=unknown
  pin=5   SMA2/U.FL2           parentID=1  dir=input  prio=2    admin=selectable   oper=unknown
  pin=6   GNSS-1PPS            parentID=0  dir=input  prio=255  admin=selectable   oper=unknown
  pin=6   GNSS-1PPS            parentID=1  dir=input  prio=255  admin=selectable   oper=unknown
  pin=7   REF-SMA1             parentID=1  dir=output prio=n/a  admin=connected    oper=unknown
  pin=8   REF-SMA2/U.FL2       parentID=1  dir=output prio=n/a  admin=connected    oper=unknown
  pin=9   PHY-CLK              parentID=0  dir=output prio=n/a  admin=connected    oper=unknown
  pin=10  MAC-CLK              parentID=0  dir=output prio=n/a  admin=connected    oper=unknown
  pin=11  CVL-SDP21            parentID=1  dir=output prio=n/a  admin=connected    oper=unknown
  pin=12  CVL-SDP23            parentID=1  dir=output prio=n/a  admin=connected    oper=unknown
=== end pin table 0x507c6fffff4b9950 ===
I0707 09:14:09.343290 3202930 pin_table.go:35] === DPLL pin table (ens2f0 pps->LOCKED) clockID=0x507c6fffff836b8c ===
  pin=17  CVL-SDP22            parentID=2  dir=input  prio=255  admin=selectable   oper=unknown
  pin=17  CVL-SDP22            parentID=3  dir=input  prio=0    admin=connected    oper=unknown
  pin=18  CVL-SDP20            parentID=2  dir=input  prio=255  admin=selectable   oper=unknown
  pin=18  CVL-SDP20            parentID=3  dir=input  prio=255  admin=selectable   oper=unknown
  pin=19  C827_0-RCLKA         parentID=2  dir=input  prio=8    admin=selectable   oper=unknown
  pin=19  C827_0-RCLKA         parentID=3  dir=input  prio=8    admin=selectable   oper=unknown
  pin=20  C827_0-RCLKB         parentID=2  dir=input  prio=9    admin=selectable   oper=unknown
  pin=20  C827_0-RCLKB         parentID=3  dir=input  prio=9    admin=selectable   oper=unknown
  pin=21  SMA1                 parentID=2  dir=input  prio=3    admin=selectable   oper=unknown
  pin=21  SMA1                 parentID=3  dir=input  prio=3    admin=selectable   oper=unknown
  pin=22  SMA2/U.FL2           parentID=2  dir=input  prio=2    admin=selectable   oper=unknown
  pin=22  SMA2/U.FL2           parentID=3  dir=input  prio=2    admin=selectable   oper=unknown
  pin=23  GNSS-1PPS            parentID=2  dir=input  prio=255  admin=selectable   oper=unknown
  pin=23  GNSS-1PPS            parentID=3  dir=input  prio=255  admin=selectable   oper=unknown
  pin=24  REF-SMA1             parentID=3  dir=output prio=n/a  admin=connected    oper=unknown
  pin=25  REF-SMA2/U.FL2       parentID=3  dir=output prio=n/a  admin=connected    oper=unknown
  pin=26  PHY-CLK              parentID=2  dir=output prio=n/a  admin=connected    oper=unknown
  pin=27  MAC-CLK              parentID=2  dir=output prio=n/a  admin=connected    oper=unknown
=== end pin table 0x507c6fffff836b8c ===
I0707 09:14:09.343295 3202930 pin_table.go:35] === DPLL pin table (ens2f0 pps->LOCKED) clockID=0x507c6fffff836aa0 ===
  pin=34  CVL-SDP22            parentID=4  dir=input  prio=255  admin=selectable   oper=unknown
  pin=34  CVL-SDP22            parentID=5  dir=input  prio=5    admin=selectable   oper=unknown
  pin=35  CVL-SDP20            parentID=4  dir=input  prio=255  admin=selectable   oper=unknown
  pin=35  CVL-SDP20            parentID=5  dir=input  prio=4    admin=selectable   oper=unknown
  pin=36  C827_0-RCLKA         parentID=4  dir=input  prio=8    admin=selectable   oper=unknown
  pin=36  C827_0-RCLKA         parentID=5  dir=input  prio=8    admin=selectable   oper=unknown
  pin=37  C827_0-RCLKB         parentID=4  dir=input  prio=9    admin=selectable   oper=unknown
  pin=37  C827_0-RCLKB         parentID=5  dir=input  prio=9    admin=selectable   oper=unknown
  pin=38  SMA1                 parentID=4  dir=input  prio=3    admin=selectable   oper=unknown
  pin=38  SMA1                 parentID=5  dir=input  prio=3    admin=selectable   oper=unknown
  pin=39  SMA2/U.FL2           parentID=4  dir=input  prio=2    admin=selectable   oper=unknown
  pin=39  SMA2/U.FL2           parentID=5  dir=input  prio=2    admin=selectable   oper=unknown
  pin=40  GNSS-1PPS            parentID=4  dir=input  prio=255  admin=selectable   oper=unknown
  pin=40  GNSS-1PPS            parentID=5  dir=input  prio=255  admin=selectable   oper=unknown
  pin=41  REF-SMA1             parentID=5  dir=output prio=n/a  admin=connected    oper=unknown
  pin=42  REF-SMA2/U.FL2       parentID=5  dir=output prio=n/a  admin=connected    oper=unknown
  pin=43  PHY-CLK              parentID=4  dir=output prio=n/a  admin=connected    oper=unknown
  pin=44  MAC-CLK              parentID=4  dir=output prio=n/a  admin=connected    oper=unknown
  pin=45  CVL-SDP21            parentID=5  dir=output prio=n/a  admin=connected    oper=unknown
  pin=46  CVL-SDP23            parentID=5  dir=output prio=n/a  admin=connected    oper=unknown
=== end pin table 0x507c6fffff836aa0 ===
```

**Comment by vitus133** [comment]
> This is an example of what logs show when a DPLL event happen (FREERUN-LOCKED): @vitus133
> 
> ```
> I0707 09:14:09.343265 3202930 pin_table.go:35] === DPLL pin table (ens2f0 pps->LOCKED) clockID=0x507c6fffff4b9950 ===
>   pin=0   CVL-SDP22            parentID=0  dir=input  prio=255  admin=selectable   oper=unknown
>   pin=0   CVL-SDP22            parentID=1  dir=input  prio=5    admin=selectable   oper=unknown
>   pin=1   CVL-SDP20            parentID=0  dir=input  prio=255  admin=selectable   oper=unknown
>   pin=1   CVL-SDP20            parentID=1  dir=input  prio=4    admin=selectable   oper=unknown
>   pin=2   C827_0-RCLKA         parentID=0  dir=input  prio=8    admin=selectable   oper=unknown
>   pin=2   C827_0-RCLKA         parentID=1  dir=input  prio=8    admin=selectable   oper=unknown
>   pin=3   C827_0-RCLKB         parentID=0  dir=input  prio=9    admin=selectable   oper=unknown
>   pin=3   C827_0-RCLKB         parentID=1  dir=input  prio=9    admin=selectable   oper=unknown
>   pin=4   SMA1                 parentID=0  dir=input  prio=3    admin=selectable   oper=unknown
>   pin=4   SMA1                 parentID=1  dir=input  prio=3    admin=selectable   oper=unknown
>   pin=5   SMA2/U.FL2           parentID=0  dir=input  prio=2    admin=selectable   oper=unknown
>   pin=5   SMA2/U.FL2           parentID=1  dir=input  prio=2    admin=selectable   oper=unknown
>   pin=6   GNSS-1PPS            parentID=0  dir=input  prio=255  admin=selectable   oper=unknown
>   pin=6   GNSS-1PPS            parentID=1  dir=input  prio=255  admin=selectable   oper=unknown
>   pin=7   REF-SMA1             parentID=1  dir=output prio=n/a  admin=connected    oper=unknown
>   pin=8   REF-SMA2/U.FL2       parentID=1  dir=output prio=n/a  admin=connected    oper=unknown
>   pin=9   PHY-CLK              parentID=0  dir=output prio=n/a  admin=connected    oper=unknown
>   pin=10  MAC-CLK              parentID=0  dir=output prio=n/a  admin=connected    oper=unknown
>   pin=11  CVL-SDP21            parentID=1  dir=output prio=n/a  admin=connected    oper=unknown
>   pin=12  CVL-SDP23            parentID=1  dir=output prio=n/a  admin=connected    oper=unknown
> === end pin table 0x507c6fffff4b9950 ===
> I0707 09:14:09.343290 3202930 pin_table.go:35] === DPLL pin table (ens2f0 pps->LOCKED) clockID=0x507c6fffff836b8c ===
>   pin=17  CVL-SDP22            parentID=2  dir=input  prio=255  admin=selectable   oper=unknown
>   pin=17  CVL-SDP22            parentID=3  dir=input  prio=0    admin=connected    oper=unknown
>   pin=18  CVL-SDP20            parentID=2  dir=input  prio=255  admin=selectable   oper=unknown
>   pin=18  CVL-SDP20            parentID=3  dir=input  prio=255  admin=selectable   oper=unknown
>   pin=19  C827_0-RCLKA         parentID=2  dir=input  prio=8    admin=selectable   oper=unknown
>   pin=19  C827_0-RCLKA         parentID=3  dir=input  prio=8    admin=selectable   oper=unknown
>   pin=20  C827_0-RCLKB         parentID=2  dir=input  prio=9    admin=selectable   oper=unknown
>   pin=20  C827_0-RCLKB         parentID=3  dir=input  prio=9    admin=selectable   oper=unknown
>   pin=21  SMA1                 parentID=2  dir=input  prio=3    admin=selectable   oper=unknown
>   pin=21  SMA1                 parentID=3  dir=input  prio=3    admin=selectable   oper=unknown
>   pin=22  SMA2/U.FL2           parentID=2  dir=input  prio=2    admin=selectable   oper=unknown
>   pin=22  SMA2/U.FL2           parentID=3  dir=input  prio=2    admin=selectable   oper=unknown
>   pin=23  GNSS-1PPS            parentID=2  dir=input  prio=255  admin=selectable   oper=unknown
>   pin=23  GNSS-1PPS            parentID=3  dir=input  prio=255  admin=selectable   oper=unknown
>   pin=24  REF-SMA1             parentID=3  dir=output prio=n/a  admin=connected    oper=unknown
>   pin=25  REF-SMA2/U.FL2       parentID=3  dir=output prio=n/a  admin=connected    oper=unknown
>   pin=26  PHY-CLK              parentID=2  dir=output prio=n/a  admin=connected    oper=unknown
>   pin=27  MAC-CLK              parentID=2  dir=output prio=n/a  admin=connected    oper=unknown
> === end pin table 0x507c6fffff836b8c ===
> I0707 09:14:09.343295 3202930 pin_table.go:35] === DPLL pin table (ens2f0 pps->LOCKED) clockID=0x507c6fffff836aa0 ===
>   pin=34  CVL-SDP22            parentID=4  dir=input  prio=255  admin=selectable   oper=unknown
>   pin=34  CVL-SDP22            parentID=5  dir=input  prio=5    admin=selectable   oper=unknown
>   pin=35  CVL-SDP20            parentID=4  dir=input  prio=255  admin=selectable   oper=unknown
>   pin=35  CVL-SDP20            parentID=5  dir=input  prio=4    admin=selectable   oper=unknown
>   pin=36  C827_0-RCLKA         parentID=4  dir=input  prio=8    admin=selectable   oper=unknown
>   pin=36  C827_0-RCLKA         parentID=5  dir=input  prio=8    admin=selectable   oper=unknown
>   pin=37  C827_0-RCLKB         parentID=4  dir=input  prio=9    admin=selectable   oper=unknown
>   pin=37  C827_0-RCLKB         parentID=5  dir=input  prio=9    admin=selectable   oper=unknown
>   pin=38  SMA1                 parentID=4  dir=input  prio=3    admin=selectable   oper=unknown
>   pin=38  SMA1                 parentID=5  dir=input  prio=3    admin=selectable   oper=unknown
>   pin=39  SMA2/U.FL2           parentID=4  dir=input  prio=2    admin=selectable   oper=unknown
>   pin=39  SMA2/U.FL2           parentID=5  dir=input  prio=2    admin=selectable   oper=unknown
>   pin=40  GNSS-1PPS            parentID=4  dir=input  prio=255  admin=selectable   oper=unknown
>   pin=40  GNSS-1PPS            parentID=5  dir=input  prio=255  admin=selectable   oper=unknown
>   pin=41  REF-SMA1             parentID=5  dir=output prio=n/a  admin=connected    oper=unknown
>   pin=42  REF-SMA2/U.FL2       parentID=5  dir=output prio=n/a  admin=connected    oper=unknown
>   pin=43  PHY-CLK              parentID=4  dir=output prio=n/a  admin=connected    oper=unknown
>   pin=44  MAC-CLK              parentID=4  dir=output prio=n/a  admin=connected    oper=unknown
>   pin=45  CVL-SDP21            parentID=5  dir=output prio=n/a  admin=connected    oper=unknown
>   pin=46  CVL-SDP23            parentID=5  dir=output prio=n/a  admin=connected    oper=unknown
> === end pin table 0x507c6fffff836aa0 ===
> ```

Yep, that really is too much for a log. First of all, when one of DPLLs reports its state changed, we don't want pins of all the other DPLLs, because they are not relevant to the change. Let's filter it by the clock ID that triggered the notification. Second, can we put one pin per line, not per parent? There is no way one parent would be input, and another parent of the same pin would be an output, so no need to have two lines per pin. 
The lines can be much shorter if you print a table header and don't print the keys in lines (such as parentID, dir, prio, admin, oper etc).
For example:
```
id     Pkg. L    prio0 adm                    oper              prio1 adm.               oper.           Brd. L
--------------------------------------------------------------------------------------------------------
16    REF0P    14      disconnected    standby        6       selectable      active         ETH01_SDP_TIMESYNC_2
17    REF0N    14      disconnected    standby        7       selectable      standby        ETH01_SDP_TIMESYNC_0
18    REF1P    10      selectable      qual-failed    10      selectable      qual-failed    CLK_78M125_NAC0_SYNCE0
19    REF1N    11      selectable      qual-failed    11      selectable      qual-failed    CLK_78M125_NAC0_SYNCE1
20    REF2N    14      disconnected    qual-failed    14      disconnected    qual-failed    GNSS_10M_IN
21    REF3P    3       selectable      qual-failed    3       selectable      qual-failed    1EPPS_IN
22    REF3N    2       selectable      qual-failed    2       selectable      qual-failed    null
25    OUT3     null    disconnected    null           null    disconnected    null           156M25_NAC0_CLKREF_SYNC
31    NCO      null    disconnected    standby        null    disconnected    standby        null
```
And the last note - we don't care about the output pins, as they are not related to the device notification that triggered this print. When we change output - it changes, and it's printed right away.

There is also another idea: EEC DPLL and PPS DPLL send indications separately, so we can print their correspondent parent devices with each DPLL device, reducing the table volume. For example, when EEC DPLL reports state change, print EEC parents, and the same with PPS. 
I'm concerned about the log volume so maybe we can filter the pins per device state. For Unlocked and Holdover, print all the enabled inputs as now. For Locked and LHAQ - print only the active input

For formatting you can use a table renderer and then print its output with glog:
```
package main

import (
	"bytes"
	"flag"

	"github.com/golang/glog"
	"github.com/rodaine/table"
)

func main() {
	// glog requires flags to be parsed before logging
	flag.Parse()
	defer glog.Flush()

	// 1. Create a buffer to capture the table output
	var buf bytes.Buffer

	// 2. Initialize the table and set the buffer as its writer
	tbl := table.New("ID", "Name", "Role")
	tbl.WithWriter(&buf)

	// 3. Add rows to the table
	tbl.AddRow(1, "Alice", "Admin")
	tbl.AddRow(2, "Bob", "User")
	tbl.AddRow(3, "Charlie", "Developer")

	// 4. "Print" the table (this writes to the bytes.Buffer instead of stdout)
	tbl.Print()

	// 5. Log the buffer contents using glog
	// Note: We add a newline (\n) before the buffer string. 
	// This is important because glog adds a timestamp/level prefix to the first line, 
	// which would misalign the first row of your ASCII table if you don't push it down.
	glog.Infof("User Data:\n%s", buf.String())
}
```
(Code snippet from Gemini)

When filtering the active pin, you can use or get inspiration from here:
https://github.com/k8snetworkplumbingwg/linuxptp-daemon/blob/db889bd1d8e93127c50f73b7874e11d4adca8713/pkg/dpll/dpll.go#L343

File: pkg/dpll-netlink/pin_table.go [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+	var buf bytes.Buffer
+	tbl := table.New("id", "Brd. L", "Pkg. L", "prio", "adm.", "oper.")
+	tbl.WithWriter(&buf)
+	for _, r := range rows {
+		tbl.AddRow(r.id, r.boardLabel, r.packageLabel, r.prio, r.admin, r.oper)
+	}
+	tbl.Print()
+
+	glog.Infof("=== DPLL pin table (%s) clockID=%#x ===\n%s", reason, clockID, buf.String())
+}
```
[the remark this answers is not in the document]
Comment (gtannous-spec) [reply pkg/dpll-netlink/pin_table.go:301]: fixed.

File: addons/intel/clock-chain.go [hunk 8]
```diff
@@ -621,17 +620,14 @@ func batchPinSet(commands []dpll.PinParentDeviceCtl) error {
...
 			glog.Error("failed to get pin: ", err)
 			return err
 		}
-		reply, err := dpll.GetPinInfoHR(info, time.Now())
-		if err != nil {
-			glog.Error("failed to convert pin reply to human readable: ", err)
-			return err
-		}
-		glog.Info("pin reply: ", string(reply))
 	}
```
Comment (vitus133) [inline addons/intel/clock-chain.go]: We want to keep printing the specific pin here as well, from two reasons: 
1. we don't print output pins elsewhere 
2. For any pin, we want to see how the pin configuration looks like after we applied the changes. 
It will be different from what is printed upon DPLL state change. For example when you enable an input pin, you will see it here as `selectable`, and when DPLL locks it will be `connected`.
Just use the table format here as well

File: pkg/hardwareconfig/utils.go
Comment (vitus133) [inline pkg/hardwareconfig/utils.go:493]: Same here. We want the specific pin printed as a table row here as well

File: addons/intel/clock-chain.go
Comment (vitus133) [inline addons/intel/clock-chain.go]: Can func BatchPinSet(commands []dpll.PinParentDeviceCtl) from hardwareconfig package be reused here?

File: pkg/dpll-netlink/pin_table.go [hunk 9]
```diff
@@ -0,0 +1,224 @@
...
+// LogPinTable logs a compact table of the input pins relevant to the DPLL
+// device (deviceID) on the given clockID that just reported a lock-state
+// change (reason is a free-form description used in the log line, e.g.
+// "eth0 eec->LOCKED").
+//
+// This is a best-effort diagnostic: the actual netlink dial/dump happens in a
+// background goroutine so callers on latency-sensitive paths (e.g.
+// nlUpdateState, processing notifications inline) are never blocked or
+// stalled by it.
+func LogPinTable(reason string, clockID uint64, deviceID uint32, lockStatus uint32) {
```
Comment (vitus133) [inline pkg/dpll-netlink/pin_table.go:255]: I think it does much more than just logging the table (the name does not suggest that). It dials, dumps, converts to table and prints. 
In general I think we need this pin_table collection to export functions to consistently print a list of pins (or a list of one pin) from another package. Maybe a function that receives a list of pins (pins []*PinInfo) and pins a table should be exported and reused from other modules?

**Review by vitus133 — COMMENTED** [review]
Added more suggestions

File: pkg/dpll-netlink/pin_table.go (hunk 9 above)
Comment (gtannous-spec) [reply pkg/dpll-netlink/pin_table.go:255]: > I think it does much more than just logging the table (the name does not suggest that). It dials, dumps, converts to table and prints.

Yes that's a fix from codeRabbit review to avoid a fresh netlink dial in the notification path.

File: addons/intel/clock-chain.go
Comment (gtannous-spec) [reply addons/intel/clock-chain.go]: I guess the logic could be exported to addons/intel/clock-chain.go ,
but the question is:  are there any dependency cycle risks?

File: pkg/dpll-netlink/pin_table.go (hunk 9 above)
Comment (gtannous-spec) [reply pkg/dpll-netlink/pin_table.go:255]: Fixed !! PTAL @vitus133

File: addons/intel/clock-chain.go
Comment (gtannous-spec) [reply addons/intel/clock-chain.go]: Fixed ..

**Review by vitus133 — APPROVED** [review]
